core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -711,6 +723,9 @@ private[log] class Cleaner(val id: Int,
           shallowOffsetOfMaxTimestamp = result.shallowOffsetOfMaxTimestamp,
           records = retained)
         throttler.maybeThrottle(outputBuffer.limit())
+        if (newCanUpdateBaseOffset)
+          dest.updateBaseOffset(result.minOffset())
+        newCanUpdateBaseOffset = false
```

Comment:
Can we move this code to `LogSegment::append`? Meaning shouldn't the `baseOffset` of a segment always be the `baseOffset` of the first record batch appended?

Is the issue that the new clean segment is created before the log cleaner knows the first clean record (record batch). If so, why don't we delay creating of the new segment until that point?

Reply:
Yeah. I'm not totally happy with how this is handled, but I'm not sure if there is another better way. So the new cleaned segment is created in cleanSegments which calls this method. The issue is that multiple segments can be compacted into a single segment and each time a segment is compacted this method is called. Thus, this method does not always know if there is a record in the new segment yet.  This is why we have this extra boolean variable. 

This is also why delaying might not work without a variable like this. The method will not know if the new segment has been created already or not.

Reply:
Okay. We can do this in a future PR if you like. Do you want to file an issue for this?

Reply:
I think one solution is to change the signature to:

```scala
private[log] def cleanInto(
    log: Log
    sourceRecords: FileRecords,
    dest: Option[LogSegment],
    map: OffsetMap,
    retainDeletesAndTxnMarkers: Boolean,
    maxLogMessageSize: Int,
    transactionMetadata: CleanedTransactionMetadata,
    lastRecordsOfActiveProducers: Map[Long, LastRecord],
    stats: CleanerStats): Optional[LogSegment]
```

This should allow you to call `LogCleaner.createNewCleanedSegment(log, result.minOffset)` when a cleaned batch has been identified. What do you think?

Reply:
This is much more elegant! A question I would have is what we do if we return from trying all the segments and we end up never making a new segment. Currently the behavior is to return an empty segment with the original baseOffset. I'd lean towards keeping the existing logic, but I suppose we could also simply use deleteSegments instead of replaceSegments in this case too.

Reply:
I've updated to delete segments that end up being empty.